src/mlpack/methods/pca/decomposition_policies/stochastic_method.hpp

```diff
@@ -0,0 +1,299 @@
+
+    // Approximate eigenvalues and eigenvectors using Rayleigh–Ritz method.
+    arma::mat u, v;
+    arma::svd_econ(u, eigVal, v, transformedData);
+
+    // Now we must square the singular values to get the eigenvalues.
+    // In addition we must divide by the number of points, because the
+    // covariance matrix is X * X' / (N - 1).
+    eigVal %= eigVal / (data.n_cols - 1);
+  }
+
+  //! Get the optimizer.
+  OptimizerType Optimizer() const { return optimizer; }
+  //! Modify the optimizer.
+  OptimizerType& Optimizer() { return optimizer; }
+
+ private:
+  //! Adjust the step size if required.
+  template<typename T>
+  inline typename std::enable_if<
+      HasStepSize<T, double&(T::*)()>::value, void>::type
+  SetStepSize(T& optimizer)
+  {
+    if (optimizer.StepSize() > 0)
+    {
+      Log::Warn << "StepSize(): invalid value (> 0), automatically take the "
+          << "negative direction." << std::endl;
+      optimizer.StepSize() *= -1;
+    }
+  }
+
+  //! Adjust the step size if required.
+  template<typename T>
+  inline typename std::enable_if<
+      !HasStepSize<T, double&(T::*)()>::value, void>::type
+  SetStepSize(T& /* optimizer */)
+  {
+    /* Nothing to do here */
+  }
+
+  OptimizerType optimizer;
+
+  /**
+   * The principal component analysis (PCA) function for the
+   * Principal component analysis (PCA). This is used by various mlpack
+   * optimizers to be used for principal component analysis.
+   */
+  class PCAFunction
+  {
+   public:
+    /**
+     * Construct The principal component analysis (PCA) function with the
+     * given data.
+     *
+     * @param data Data matrix.
+     */
+    PCAFunction(const arma::mat& data) :
+        // We promise to be well-behaved... the elements won't be modified.
+        data(math::MakeAlias(const_cast<arma::mat&>(data), false)),
+        pl(0)
+    {
+      // Nothing to do here.
+    }
+
+    /**
+     * Shuffle the order of function visitation. This may be called by the
+     * optimizer.
+     */
+    void Shuffle()
+    {
+      // Generate ordering.
+      arma::uvec ordering = arma::shuffle(arma::linspace<arma::uvec>(0,
+          data.n_cols - 1, data.n_cols));
+      arma::mat newData = data.cols(ordering);
+
+      // If we are an alias, make sure we don't write to the original data.
+      math::ClearAlias(data);
+
+      // Take ownership of the new data.
+      data = std::move(newData);
+    }
+
+    //! Return the number of functions.
+    size_t NumFunctions() const { return data.n_cols; }
+
+    /*
+     * Evaluate a function for a particular batch-size.
+     *
+     * @param coordinates The function coordinates.
+     * @param begin The first function.
+     * @param batchSize Number of points to process.
+     */
+    double Evaluate(const arma::mat& /* coordinates */,
+                    const size_t /* begin */,
+                    const size_t /* batchSize */)
+    {
+      return pl;
```

Comment:
I'm a bit confused about the psuedoloss here... I think you are trying to avoid the computation of the objective entirely (so maybe my comment about negative the objective makes no sense, since I wrote it before I got here!), but I'd be concerned that the optimizer might terminate early because it detects convergence.

From equation 1, I guess that the objective is just `tr(U^T x x^T U)`.  If we have some `EvaluateWithGradient()`, then we only have to do the extra computation of pre-multiplying `U^T` to the term we take in the gradient, and then taking the trace of that.  So maybe it is not that much extra, but I could see that it might still make a difference.

What do you think, is there any problem computing the actual loss?  I am not opposed to avoiding the objective calculation if it's not any problem.